contrib/redis/go-redis.v9/redis.go

```diff
@@ -0,0 +1,183 @@
+// Unless explicitly stated otherwise all files in this repository are licensed
+// under the Apache License Version 2.0.
+// This product includes software developed at Datadog (https://www.datadoghq.com/).
+// Copyright 2016 Datadog, Inc.
+
+// Package redis provides functions to trace the redis/go-redis package (https://github.com/redis/go-redis).
+package redis
+
+import (
+	"bytes"
+	"context"
+	"math"
+	"net"
+	"strconv"
+	"strings"
+
+	"gopkg.in/DataDog/dd-trace-go.v1/ddtrace"
+	"gopkg.in/DataDog/dd-trace-go.v1/ddtrace/ext"
+	"gopkg.in/DataDog/dd-trace-go.v1/ddtrace/tracer"
+
+	"github.com/redis/go-redis/v9"
+)
+
+type datadogHook struct {
+	*params
+}
+
+// params holds the tracer and a set of parameters which are recorded with every trace.
+type params struct {
+	config         *clientConfig
+	additionalTags []ddtrace.StartSpanOption
+}
+
+// NewClient returns a new Client that is traced with the default tracer under
+// the service name "redis".
+func NewClient(opt *redis.Options, opts ...ClientOption) redis.UniversalClient {
+	client := redis.NewClient(opt)
+	WrapClient(client, opts...)
+	return client
+}
+
+// WrapClient adds a hook to the given client that traces with the default tracer under
+// the service name "redis".
+func WrapClient(client redis.UniversalClient, opts ...ClientOption) {
+	cfg := new(clientConfig)
+	defaults(cfg)
+	for _, fn := range opts {
+		fn(cfg)
+	}
+
+	hookParams := &params{
+		additionalTags: additionalTagOptions(client),
+		config:         cfg,
+	}
+
+	client.AddHook(&datadogHook{params: hookParams})
+}
+
+type clientOptions interface {
+	Options() *redis.Options
+}
+
+type clusterOptions interface {
+	Options() *redis.ClusterOptions
+}
+
+func additionalTagOptions(client redis.UniversalClient) []ddtrace.StartSpanOption {
+	additionalTags := []ddtrace.StartSpanOption{}
+	if clientOptions, ok := client.(clientOptions); ok {
+		opt := clientOptions.Options()
+		if opt.Addr == "FailoverClient" {
+			additionalTags = []ddtrace.StartSpanOption{
+				tracer.Tag("out.db", strconv.Itoa(opt.DB)),
+			}
+		} else {
+			host, port, err := net.SplitHostPort(opt.Addr)
+			if err != nil {
+				host = opt.Addr
+				port = "6379"
+			}
+			additionalTags = []ddtrace.StartSpanOption{
+				tracer.Tag(ext.TargetHost, host),
+				tracer.Tag(ext.TargetPort, port),
+				tracer.Tag("out.db", strconv.Itoa(opt.DB)),
+			}
+		}
+	} else if clientOptions, ok := client.(clusterOptions); ok {
+		addrs := []string{}
+		for _, addr := range clientOptions.Options().Addrs {
+			addrs = append(addrs, addr)
+		}
+		additionalTags = []ddtrace.StartSpanOption{
+			tracer.Tag("addrs", strings.Join(addrs, ", ")),
+		}
+	}
+	return additionalTags
+}
+
+func (ddh *datadogHook) DialHook(hook redis.DialHook) redis.DialHook {
+	return func(ctx context.Context, network, addr string) (net.Conn, error) {
+		conn, err := hook(ctx, network, addr)
+		return conn, err
```

Comment:
We also wanna be able to trace the dial connection since now it's supported by the library (please make sure to also add tests for this).

You can follow the same structure as the span created in `ProcessHook`, but just using the common tags from `ddh.additionalTags` and making sure we respect options like `ddh.params.config.analyticsRate`. We can tentatively name the span `redis.dial` (pending confirmation).

Also please let me know if I can provide some further help/guidance with the implementation!

Reply:
I am in the process of implementing spans for dial.
The additional dial spans will change the total number of spans, which will affect many existing test cases and will require more modifications.
It may take some time, but I will push it as soon as it is ready. Thank you.

Reply:
Hello, @rarguelloF,

I've implemented the dial connection tracing and updated the affected tests in [b26b374](https://github.com/DataDog/dd-trace-go/pull/1730/commits/b26b374763474fd792f2f5d94d55f0ec5b4c9b0b).
Here are the key changes I made:

in redis.go:
- Added `serviceName`, `ddh.additionalTags` and `analyticsRate` tags
- Set `"redis.dial"` as the `operationName`. 

in redis_test.go:
- Explicitly picked up the `redis.command` span to handle the change in the total number of spans
- Added a new TestDial function to explicitly pick up the `redis.dial` span and check the tags

Please review the changes [b26b374](https://github.com/DataDog/dd-trace-go/pull/1730/commits/b26b374763474fd792f2f5d94d55f0ec5b4c9b0b) and let me know if you have any further suggestions or questions.

Thank you for your guidance and support.